### ISSUE TITLE
DEALER <-> ROUTER basic async message handling

### ISSUE THREAD

**Issue by RangelAngelov** [zeromq/jeromq#852]
Hello team,

Some guidance would be highly appreciated.
My situation is the following:
I have a 1 DEALER and 1 ROUTER socket.

DEALER code:
mainLoop() {
 dealer.setReceiveTimeOut(0);
 while(!queue.isEmpty()) {
                    var event = queue.poll();
                    if(Objects.nonNull(event)) {
                        dealer.send(event.toString());
                    }
                }
                byte[] messageBytes = dealer.recv();
                if (Objects.isNull(messageBytes)) {
                    continue;
                }
               //handle message...
}

The ROUTER socket code is not all that important.
It basically handles events and sends back some response message to the DEALER.

As can be seen, the DEALER makes use of an event queue to propagate events to the ROUTER.
The aim is to send multiple events to the ROUTER without having to block recv() and wait for a response for each one.

The problem with the above code is that when idle, the DEALER socket repeatedly calls recv() and loops again, which
causes it to consume a lot of CPU time.

A workaround is to set receive timeout to some value N in order to cause the DEALER to block for a little while every time
recv() is called and the inbound message queue is empty.
However this causes another problem which leads to an event being sent for processing with an N milliseconds delay.
Which is not ideal, we aim to minimise the amount of time an event is left in the event queue.

Any tips on how to avoid consuming crazy amounts of CPU for being idle and still maintaining event propagation without delay would be highly appreciated.

**Comment by fredoboulo** [comment]
Hello,

there could be multiple ways to address that. My individual taste would go towards using a ZActor, like in [ZMonitor](https://github.com/zeromq/jeromq/blob/master/src/main/java/org/zeromq/ZMonitor.java)

The pattern would be a main actor thread dealing with I/O operations (like your send and receive ops), with a command pipe that can send messages from another thread. You would still need the queue somehow, as this command pipe is not thread-safe (like every sockets).

At a much lower level, it makes use of pollers to be more efficient and non-blocking (like in http://zguide.zeromq.org/java:mspoller). [This section](Multithreading with ZeroMQ) could also be useful.

**Comment by RangelAngelov** [comment]
Hello,

I'd imagine that the command pipe implementation would be PUSH-PULL inproc sockets, which would allow us to propagate message handling to other threads, this definitely sounds handy. I also realise the persistent need for the eventQueue, due to the ZMQ sockets not being thread safe.
However I am having a hard time grasping the concept behind the ZActor and how it would allow for async event sending/handling.
Would you care to elaborate on the aforementioned subject.

Thanks

**Comment by RangelAngelov** [comment]
Hello,

Thank you very much for the exhaustive answer.
I eventually ended up using two inproc PUSH-PULL socket pairs for one-directional feeding of outbound requests and inbound responses into the main processing loop.
I also made use of a poller in the main loop in order to prioritise sending requests over handling responses and block when there are no zmq events queued for processing.

This approach seems to be an alright solution for now, as the use case is not all that intricate.
However the actor pattern implementation has definitely caught my interest (didn't know about it until now) and I most likely will opt into using it eventually.

Im proceeding to closing this ticket as the case is resolved.

Thanks
